modbus: serial : Adding support for serial async

jalinei opened the pull request:
Async serial support is implemented by adding serial async functions in modbus_serial and by adding a choice for the serial driver in Kconfig. 
Tested on nucleo_g474re using rtu_server and rtu_client modbus samples. I've ran the sample up to 10625000 Baud reliably.

Fixes #52568

Signed-off-by: Jean Alinei <jean.alinei@laas.fr>

jfischer-no reviewed (changes requested):
- remove "Fixes #52568" from the commit message, git does not know anything about github
- fix style issues, follow project and subsystem coding style, take care of preferred line length
- that we have multiple APIs for UART drivers is a misery. Extending this to other subsystems is problematic.

jalinei committed:
modbus: serial : Adding support for serial async

Implemented by adding serial async functions
in modbus_serial and a choice for the serial
driver in Kconfig.

Signed-off-by: Jean Alinei <jean.alinei@laas.fr>

jfischer-no commented on subsys/modbus/Kconfig:
```diff
@@ -47,6 +47,29 @@ config MODBUS_SERIAL
 	help
 	  Enable Modbus over serial line support.
 
+choice
+
+	prompt "Supported serial drivers"
+	default MODBUS_SERIAL_IRQ
+	help
+	  Specify which serial driver to be used.
+
+config MODBUS_SERIAL_IRQ
+	bool "Modbus over serial line using IRQ uart driver"
+	depends on SERIAL && SERIAL_HAS_DRIVER
+	depends on DT_HAS_ZEPHYR_MODBUS_SERIAL_ENABLED
+	help
+	  Enable Modbus over serial line support using IRQ method.
+
+config MODBUS_SERIAL_ASYNC
+	bool "Modbus over serial line using Async uart driver"
+	depends on SERIAL && SERIAL_HAS_DRIVER && UART_ASYNC_API
+	depends on DT_HAS_ZEPHYR_MODBUS_SERIAL_ENABLED
+	help
+	  Enable Modbus over serial line support using ASYNC method.
+
+endchoice
+
```
NACK to new Kconfig options, I do not see the need for them.

jalinei replied:
I would like to understand this point better : 
Let's imagine the case where we have 2 UART peripherals.  
Only one use the Modbus protocol on top of it.  
The idea I had is that the regular UART could use async while we would still like to have the second UART with Modbus using interrupts or vice versa. Without this addition I guess both UART would be force to use the same API. 
Am I mistaking ?

jfischer-no replied:
From modbus perspective it does not matter how another UART interface is configure or used. If it has advantages to use ASYNC API (looks like only RTU would make sense), then it should be preferred when an UART driver supports it, if not, modbus serial support should fall back to IRQ API, always, no options, simple code.

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -37,14 +37,31 @@ static void modbus_serial_tx_on(struct modbus_context *ctx)
 		gpio_pin_set(cfg->de->port, cfg->de->pin, 1);
 	}
 
+#if defined CONFIG_MODBUS_SERIAL_IRQ
 	uart_irq_tx_enable(cfg->dev);
+#endif
+
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
```
NACK to this kind of ifdefery.

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -37,14 +37,31 @@ static void modbus_serial_tx_on(struct modbus_context *ctx)
+#endif
+
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
+	int ret;
+	uint8_t *buf = cfg->uart_buf;
+	ret = uart_tx(cfg->dev, buf, cfg->uart_buf_ctr, SYS_FOREVER_US);
+
+	if (ret) {
+		LOG_ERR("uart_tx() failed, err %d", ret);
+	}
+	LOG_HEXDUMP_DBG(cfg->uart_buf_ptr, cfg->uart_buf_ctr, "TX data sent:");
+	cfg->uart_buf_ptr += cfg->uart_buf_ctr;
+	cfg->uart_buf_ctr = 0;
```
Please move it to modbus_serial_async_start(struct modbus_context *ctx).
modbus_serial_tx_on() should fall back to uart_irq_tx_enable() if CONFIG_UART_ASYNC_API is not enabled.
```c

if (IS_ENABLED(CONFIG_LOG_BACKEND_UART_ASYNC)) {
    modbus_serial_async_start(ctx);
} else {
    uart_irq_tx_enable(cfg->dev);
}
```

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -58,14 +75,26 @@ static void modbus_serial_rx_on(struct modbus_context *ctx)
 		gpio_pin_set(cfg->re->port, cfg->re->pin, 1);
 	}
 
+#if defined CONFIG_MODBUS_SERIAL_IRQ
 	uart_irq_rx_enable(cfg->dev);
+#endif
+
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
+	uart_rx_enable(cfg->dev, cfg->uart_buf, sizeof(cfg->uart_buf), cfg->rtu_timeout);
+#endif
```
```c
if (IS_ENABLED(CONFIG_LOG_BACKEND_UART_ASYNC)) {
    uart_rx_enable(cfg->dev, cfg->uart_buf, sizeof(cfg->uart_buf),
                   cfg->rtu_timeout);
} else {
    uart_irq_rx_enable(cfg->dev);
}
```

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -58,14 +75,26 @@ static void modbus_serial_rx_on(struct modbus_context *ctx)
 {
 	struct modbus_serial_config *cfg = ctx->cfg;
 
+#if defined CONFIG_MODBUS_SERIAL_IRQ
 	uart_irq_rx_disable(cfg->dev);
+#endif
+
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
+	uart_rx_disable(cfg->dev);
+#endif
```
as above

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -568,7 +648,7 @@ int modbus_serial_init(struct modbus_context *ctx,
 		cfg->rtu_timeout = (numof_bits * if_delay_max) /
 				   param.serial.baud;
 	} else {
-		cfg->rtu_timeout = (numof_bits * if_delay_max) / 38400;
```
Why and how is it relevant to your changes?

jalinei replied:
Stupid mistake for that one, I will suppress this unnecessary change.

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -499,6 +530,55 @@ int modbus_serial_tx_adu(struct modbus_context *ctx)
+		break;
+	case UART_RX_RDY:
+		cfg->uart_buf_ctr += evt->data.rx.len;
+		k_work_submit(&ctx->server_work);
```
How does it handle ASCII mode?

jalinei replied:
mmh, at the current state of this PR, ASCII is not supported in async mode. 
I guess it could be added using similar if statements as in cb_handler_rx and cb_handler_tx inside of the async callback cases ? 
What is your view ?

jfischer-no replied:
I do not see any benefit to use ASYNC API for ASCII mode. I think if interface enables ASCII mode then modbus serial support should fallback to IRQ API.

jfischer-no reviewed (changes requested):
Please work away ifdefery, keep coding style and preferred line length, your code is not built and not tested, testing platform is frdm_k64f (two UARTs available) but it does not support ASYNC mode (like many others).
